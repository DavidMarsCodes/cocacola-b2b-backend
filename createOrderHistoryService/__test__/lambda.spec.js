const handler = require('../lambda');

describe('Lambda Function', () => {
  it('You should save orders in the dynamoDB database.', async () => {

    const event = {
      transactionId: '1715ba2d-5b0b-4ba0-883b-c48a0de8a670',
      cpgId: '001',
      countryId: 'AR',
      organizationId: '3043',
      data: [ {
        erpOrderId: '200111',
        erpOrderDatetime: '2020-12-31T00:00:00.000Z',
        erpOrderDeliveryDate: '2021-01-04T00:00:00.000Z',
        erpOrderStatus: 'created',
        erpOrderAmount: 1000,
        erpChannelInput: 'SFCoke',
        orderId: null,
        erpDocumentType: 'FC',
        erpPaymentMethod: 'Contado',
        erpShippingInformation: 'no aplica',
      } ],
      createdTime: expect.any(String),
    };

    const awsRepositories = { Repository: { create: jest.fn().mockResolvedValue({ ok: true }) } };

    const experts = { orderHistory: { validatecreateOrderHistory: jest.fn() } };

    const getTableName = jest.fn().mockResolvedValue({ tableName: 'OrderHistoryToUpsert' });

    const res = { sendStatus: jest.fn().mockReturnValue({ httpStatus: 201 }) };

    const { createOrderHistory } = handler({ awsRepositories, getTableName, experts, res });
    await createOrderHistory(event);

    expect(experts.orderHistory.validatecreateOrderHistory.mock.calls).toMatchSnapshot([
      [
        {
          transactionId: '1715ba2d-5b0b-4ba0-883b-c48a0de8a670',
          cpgId: '001',
          countryId: 'AR',
          organizationId: '3043',
          data: [ {
            erpOrderId: '200111',
            erpOrderDatetime: '2020-12-31T00:00:00.000Z',
            erpOrderDeliveryDate: '2021-01-04T00:00:00.000Z',
            erpOrderStatus: 'created',
            erpOrderAmount: 1000,
            erpChannelInput: 'SFCoke',
            orderId: null,
            erpDocumentType: 'FC',
            erpPaymentMethod: 'Contado',
            erpShippingInformation: 'no aplica',
          } ],
          createdTime: expect.any(String),
        },
      ],
    ]);

    expect(getTableName.mock.calls).toEqual([ [] ]);

    expect(awsRepositories.Repository.create.mock.calls).toMatchSnapshot([
      [
        'OrderHistoryToUpsert',
        {
          transactionId: '1715ba2d-5b0b-4ba0-883b-c48a0de8a670',
          cpgId: '001',
          countryId: 'AR',
          organizationId: '3043',
          data: [ {
            erpOrderId: '200111',
            erpOrderDatetime: '2020-12-31T00:00:00.000Z',
            erpOrderDeliveryDate: '2021-01-04T00:00:00.000Z',
            erpOrderStatus: 'created',
            erpOrderAmount: 1000,
            erpChannelInput: 'SFCoke',
            orderId: null,
            erpDocumentType: 'FC',
            erpPaymentMethod: 'Contado',
            erpShippingInformation: 'no aplica',
          } ],
          createdTime: expect.any(String),
        },
      ],
    ]);

    expect(res.sendStatus.mock.calls).toEqual([
      [
        201,
      ],
    ]);
  });

  it('It should answer a validation error in case any of the required parameters are invalid.', async () => {

    const event = {
      transactionId: '1715ba2d-5b0b-4ba0-883b-c48a0de8a670',
      cpgId: '001',
      countryId: 'AR',
      organizationId: '3043',
      data: [ {
        erpOrderId: '200111',
        erpOrderDatetime: '2020-12-31T00:00:00.000Z',
        erpOrderDeliveryDate: '2021-01-04T00:00:00.000Z',
        erpOrderStatus: 'created',
        erpOrderAmount: 1000,
        erpChannelInput: 'SFCoke',
        orderId: null,
        erpDocumentType: 'FC',
        erpPaymentMethod: 'Contado',
        erpShippingInformation: 'no aplica',
      } ],
      createdTime: expect.any(String),
    };

    const awsRepositories = { Repository: { create: jest.fn().mockResolvedValue({ ok: true }) } };

    const getTableName = jest.fn().mockResolvedValue({ tableName: 'OrderHistoryToUpsert' });

    const experts = {
      orderHistory: {
        validatecreateOrderHistory: jest.fn(data => {
          throw {
            customError: true,
            getData: jest.fn().mockReturnValue({
              httpStatus: 400,
              status: 'error',
              code: 100,
              msg: 'Validation_server_error',
              type: 'Validation_error',
            }),
          };
        }),
      },
    };

    const res = {
      sendStatus: jest.fn().mockReturnValue({ httpStatus: 201 }),
      error: jest.fn((message = '', code = 0, errorType = 'Internal_Server_Error', status = 500) => ({
        httpStatus: status,
        ok: false,
        code,
        errorType,
        message,
      })),
    };

    const { createOrderHistory } = handler({ awsRepositories, getTableName, experts, res });
    await createOrderHistory(event);

    expect(experts.orderHistory.validatecreateOrderHistory.mock.calls).toMatchSnapshot([
      [
        {
          transactionId: '1715ba2d-5b0b-4ba0-883b-c48a0de8a670',
          cpgId: '001',
          countryId: 'AR',
          organizationId: '3043',
          data: [ {
            erpOrderId: '200111',
            erpOrderDatetime: '2020-12-31T00:00:00.000Z',
            erpOrderDeliveryDate: '2021-01-04T00:00:00.000Z',
            erpOrderStatus: 'created',
            erpOrderAmount: 1000,
            erpChannelInput: 'SFCoke',
            orderId: null,
            erpDocumentType: 'FC',
            erpPaymentMethod: 'Contado',
            erpShippingInformation: 'no aplica',
          } ],
          createdTime: expect.any(String),
        },
      ],
    ]);

    expect(getTableName.mock.calls).toEqual([]);

    expect(awsRepositories.Repository.create.mock.calls).toEqual([]);

    expect(res.sendStatus.mock.calls).toEqual([]);

    expect(res.error.mock.calls).toEqual([
      [
        'Validation_server_error',
        100,
        'Validation_error',
        400,

      ],
    ]);

  });

});