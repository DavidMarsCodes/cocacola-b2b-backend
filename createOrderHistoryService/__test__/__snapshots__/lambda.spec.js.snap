// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lambda Function It should answer a validation error in case any of the required parameters are invalid. 1`] = `
Array [
  Array [
    Object {
      "countryId": "AR",
      "cpgId": "001",
      "createdTime": Any<String>,
      "data": Array [
        Object {
          "erpChannelInput": "SFCoke",
          "erpDocumentType": "FC",
          "erpOrderAmount": 1000,
          "erpOrderDatetime": "2020-12-31T00:00:00.000Z",
          "erpOrderDeliveryDate": "2021-01-04T00:00:00.000Z",
          "erpOrderId": "200111",
          "erpOrderStatus": "created",
          "erpPaymentMethod": "Contado",
          "erpShippingInformation": "no aplica",
          "orderId": null,
        },
      ],
      "organizationId": "3043",
      "transactionId": "1715ba2d-5b0b-4ba0-883b-c48a0de8a670",
    },
  ],
]
`;

exports[`Lambda Function You should save orders in the dynamoDB database. 1`] = `
Array [
  Array [
    Object {
      "countryId": "AR",
      "cpgId": "001",
      "createdTime": Any<String>,
      "data": Array [
        Object {
          "erpChannelInput": "SFCoke",
          "erpDocumentType": "FC",
          "erpOrderAmount": 1000,
          "erpOrderDatetime": "2020-12-31T00:00:00.000Z",
          "erpOrderDeliveryDate": "2021-01-04T00:00:00.000Z",
          "erpOrderId": "200111",
          "erpOrderStatus": "created",
          "erpPaymentMethod": "Contado",
          "erpShippingInformation": "no aplica",
          "orderId": null,
        },
      ],
      "organizationId": "3043",
      "transactionId": "1715ba2d-5b0b-4ba0-883b-c48a0de8a670",
    },
  ],
]
`;

exports[`Lambda Function You should save orders in the dynamoDB database. 2`] = `
Array [
  Array [
    "OrderHistoryToUpsert",
    Object {
      "countryId": "AR",
      "cpgId": "001",
      "createdTime": Any<String>,
      "data": Array [
        Object {
          "erpChannelInput": "SFCoke",
          "erpDocumentType": "FC",
          "erpOrderAmount": 1000,
          "erpOrderDatetime": "2020-12-31T00:00:00.000Z",
          "erpOrderDeliveryDate": "2021-01-04T00:00:00.000Z",
          "erpOrderId": "200111",
          "erpOrderStatus": "created",
          "erpPaymentMethod": "Contado",
          "erpShippingInformation": "no aplica",
          "orderId": null,
        },
      ],
      "organizationId": "3043",
      "transactionId": "1715ba2d-5b0b-4ba0-883b-c48a0de8a670",
    },
  ],
]
`;
